From c4b947d96eb273089b6753534f673640efa77703 Mon Sep 17 00:00:00 2001
From: "Gary D. Gregory" <ggregory@apache.org>
Date: Mon, 19 Mar 2012 20:56:34 +0000
Subject: [PATCH] [CODEC-130] Base64InputStream.skip skips underlying stream,
 not output. Can skip more than Integer.MAX_VALUE at a time.

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/codec/trunk@1302652 13f79535-47bb-0310-9956-ffa450edef68
---
 .../codec/binary/BaseNCodecInputStream.java      | 16 ++++++----------
 1 file changed, 6 insertions(+), 10 deletions(-)

diff --git a/src/main/java/org/apache/commons/codec/binary/BaseNCodecInputStream.java b/src/main/java/org/apache/commons/codec/binary/BaseNCodecInputStream.java
index 6c335ecfc0..78642c6e17 100644
--- a/src/main/java/org/apache/commons/codec/binary/BaseNCodecInputStream.java
+++ b/src/main/java/org/apache/commons/codec/binary/BaseNCodecInputStream.java
@@ -163,21 +163,17 @@ public long skip(long n) throws IOException {
 
         // skip in chunks of 512 bytes
         final byte[] b = new byte[512];
-        final int max = (int) Math.min(n, Integer.MAX_VALUE);
-        int total = 0;
+        long todo = n;
 
-        while (total < max) {
-            int len = max - total;
-            if (len > b.length) {
-                len = b.length;
-            }
-            len = read(b, 0, len);
+        while (todo > 0) {
+            int len = (int) Math.min(b.length, todo);
+            len = this.read(b, 0, len);
             if (len == EOF) {
                 break;
             }
-            total += len;
+            todo -= len;
         }
 
-        return total;
+        return n - todo;
     }
 }
